components/servo/lib.rs

```diff
@@ -240,11 +247,228 @@ impl<Window> Servo<Window> where Window: WindowMethods + 'static {
+                self.compositor.on_zoom_window_event(magnification);
+            }
+
+            WindowEvent::ResetZoom => {
+                self.compositor.on_zoom_reset_window_event();
+            }
+
+            WindowEvent::PinchZoom(magnification) => {
+                self.compositor.on_pinch_zoom_window_event(magnification);
+            }
+
+            WindowEvent::Navigation(top_level_browsing_context_id, direction) => {
+                let msg = ConstellationMsg::TraverseHistory(top_level_browsing_context_id, direction);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending navigation to constellation failed ({}).", e);
+                }
+            }
+
+            WindowEvent::TouchpadPressure(cursor, pressure, stage) => {
+                self.compositor.on_touchpad_pressure_event(cursor, pressure, stage);
+            }
+
+            WindowEvent::KeyEvent(ch, key, state, modifiers) => {
+                let msg = ConstellationMsg::KeyEvent(ch, key, state, modifiers);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending key event to constellation failed ({}).", e);
+                }
+            }
+
+            WindowEvent::Quit => {
+                if self.compositor.shutdown_state == ShutdownState::NotShuttingDown {
+                    debug!("Shutting down the constellation for WindowEvent::Quit");
+                    self.compositor.start_shutting_down();
+                }
+            }
+
+            WindowEvent::Reload(top_level_browsing_context_id) => {
+                let msg = ConstellationMsg::Reload(top_level_browsing_context_id);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending reload to constellation failed ({}).", e);
+                }
+            }
+
+            WindowEvent::ToggleWebRenderProfiler => {
+                let mut flags = self.compositor.webrender.get_debug_flags();
+                flags.toggle(webrender::renderer::PROFILER_DBG);
+                self.compositor.webrender.set_debug_flags(flags);
+                self.compositor.webrender_api.generate_frame(self.compositor.webrender_document, None);
+            }
+
+            WindowEvent::NewBrowser(url, response_chan) => {
+                let msg = ConstellationMsg::NewBrowser(url, response_chan);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending NewBrowser message to constellation failed ({}).", e);
+                }
+            }
+
+            WindowEvent::SelectBrowser(ctx) => {
+                let msg = ConstellationMsg::SelectBrowser(ctx);
+                if let Err(e) = self.constellation_chan.send(msg) {
+                    warn!("Sending SelectBrowser message to constellation failed ({}).", e);
+                }
+            }
+        }
+    }
+
+    fn receive_messages(&mut self) {
```

Comment:
I don't know, I was thinking that perhaps this kind of forwarding logic, while somewhat repetitive, also serves a purpose in hiding servo internals from the embedder(for example the shutdown state of the compositor), as well as maintaining a consistent API and one that is easy for an embedder to respond to(assuming it is easier for someone to implement `WindowMethods`, than it would be receiving messages).